### PR TITLE
Allows multiple connections from client to server and adds callback example for control in ROS distributed example

### PR DESCRIPTION
**Pull request by stephendb**
Simulator constructor public
Create connection for control.
Add callback to ros distributed example.
The purpose of this PR is to enable multiple threads to connect and send commands on the control channel (i.e., same port/ip).

### CODE REVIEW

File: monodrive/core/src/DistributedServer.h [hunk 1]
```diff
@@ -113,6 +113,7 @@ class DistributedServer {
     virtual bool sendCommand(ApiMessage message, nlohmann::json* response=nullptr);
     /// @brief Send command to server and return immediately
     virtual bool sendCommandAsync(ApiMessage message, nlohmann::json* response=nullptr);
+    Simulator* GetSimulator();
```
Comment (bradleynull) [inline monodrive/core/src/DistributedServer.h:116]: Can we add documentation and call this `getSimulator()` to make it consistent with the rest of the class?

File: examples/cpp/distributed/distributed.cpp [hunk 2]
```diff
@@ -32,6 +32,8 @@ void PrimaryThread(std::shared_ptr<PrimaryDistributedServer> server, Event* repl
   mono::precise_stopwatch primary_stopwatch;
 
   std::cout << "Primary thread starting..." << std::endl;
+  auto egoControlServer = std::make_unique<Simulator>(server->GetSimulator()->getServerIp(), server->GetSimulator()->getServerPort());
```
Comment (bradleynull) [inline examples/cpp/distributed/distributed.cpp:35]: Probably should add a comment on why this is necessary to help users understand when to start their own control server.

File: examples/ros/src/distributed/src/distributed.cpp [hunk 3]
```diff
@@ -203,6 +203,18 @@ int main(int argc, char **argv)
     }
   }
 
+  auto egoControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
```
Comment (bradleynull) [inline examples/ros/src/distributed/src/distributed.cpp:206]: Add a comment on why necessary

File: monodrive/core/src/Simulator.h [hunk 4]
```diff
@@ -57,9 +57,9 @@ class Simulator
 
 	const std::string& getServerIp() const{return serverIp;}
 	const short& getServerPort() const{return serverPort;}
+	Simulator(const std::string& serverIp, const short& serverPort);
```
Comment (bradleynull) [inline monodrive/core/src/Simulator.h:60]: Not needed for this stop gap, but we should add an issue that describes what needs to be done to properly protect a server from reconfiguring. You can still call `configure()` on this server instance which could mess things up. 

I think the proper solution is to have a base simulator class then one that provides configuration options via `getInstance()` and one that only provides a control connection with no `configure()` function.

**Review by bradleynull — APPROVED** [review]
@stephendb Just some minor cleanup details to clarify what is going on to the end-user.

**Review by bradleynull — CHANGES_REQUESTED** [review]
Sorry didn't mean to approve the last one.

**Review by bradleynull — CHANGES_REQUESTED** [review]
@stephendb Sorry just realized that recent updates have caused a linker error: 

```
[build]   Auto build dll exports
[build]      Creating library C:/Users/bradn/Documents/GitHub/simulator/Simulator/Plugins/monoDriveCore/Source/monodrive-client/build/monodrive/core/Debug/monodrive.lib and object C:/Users/bradn/Documents/GitHub/simulator/Simulator/Plugins/monoDriveCore/Source/monodrive-client/build/monodrive/core/Debug/monodrive.exp
[build]   monodrive.vcxproj -> C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\build\monodrive\core\Debug\monodrive.dll
[build]   Building Custom Rule C:/Users/bradn/Documents/GitHub/simulator/Simulator/Plugins/monoDriveCore/Source/monodrive-client/examples/cpp/distributed/CMakeLists.txt
[build]   Building Custom Rule C:/Users/bradn/Documents/GitHub/simulator/Simulator/Plugins/monoDriveCore/Source/monodrive-client/examples/cpp/distributed/CMakeLists.txt
[build]   distributed.cpp
[build] C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\examples\cpp\distributed\distributed.cpp(175,63): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data [C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\build\examples\cpp\distributed\distributed.vcxproj]
[build] distributed.obj : error LNK2019: unresolved external symbol "public: class Simulator * __cdecl DistributedServer::GetSimulator(void)" (?GetSimulator@DistributedServer@@QEAAPEAVSimulator@@XZ) referenced in function "void __cdecl PrimaryThread(class std::shared_ptr<class PrimaryDistributedServer>,class Event *)" (?PrimaryThread@@YAXV?$shared_ptr@VPrimaryDistributedServer@@@std@@PEAVEvent@@@Z) [C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\build\examples\cpp\distributed\distributed.vcxproj]
[build] C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\build\examples\cpp\distributed\Debug\distributed.exe : fatal error LNK1120: 1 unresolved externals [C:\Users\bradn\Documents\GitHub\simulator\Simulator\Plugins\monoDriveCore\Source\monodrive-client\build\examples\cpp\distributed\distributed.vcxproj]
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
```

**Comment by bradleynull** [comment]
@stephendb Looks like similar build issues are happening with the ros examples:

```
[ 83%] Built target wheel_vehicle_control_node
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp: In function ‘int main(int, char**)’:
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:44: error: ‘Server’ was not declared in this scope
   auto egoControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                            ^~~~~~
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:44: note: suggested alternative: ‘Sensor’
   auto egoControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                            ^~~~~~
                                            Sensor
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:144: error: no matching function for call to ‘make_unique<<expression error> >(const string&, const short int&)’
 oControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                                                                                                                       ^
In file included from /usr/include/c++/7/memory:80:0,
                 from /usr/include/c++/7/thread:39,
                 from /home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:820:5: note: candidate: template<class _Tp, class ... _Args> typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...)
     make_unique(_Args&&... __args)
     ^~~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:820:5: note:   template argument deduction/substitution failed:
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:144: error: template argument 1 is invalid
 oControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                                                                                                                       ^
In file included from /usr/include/c++/7/memory:80:0,
                 from /usr/include/c++/7/thread:39,
                 from /home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:826:5: note: candidate: template<class _Tp> typename std::_MakeUniq<_Tp>::__array std::make_unique(std::size_t)
     make_unique(size_t __num)
     ^~~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:826:5: note:   template argument deduction/substitution failed:
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:144: error: template argument 1 is invalid
 oControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                                                                                                                       ^
In file included from /usr/include/c++/7/memory:80:0,
                 from /usr/include/c++/7/thread:39,
                 from /home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:832:5: note: candidate: template<class _Tp, class ... _Args> typename std::_MakeUniq<_Tp>::__invalid_type std::make_unique(_Args&& ...) <deleted>
     make_unique(_Args&&...) = delete;
     ^~~~~~~~~~~
/usr/include/c++/7/bits/unique_ptr.h:832:5: note:   template argument deduction/substitution failed:
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:144: error: template argument 1 is invalid
 oControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
                                                                                                                                       ^
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:209:60: error: base operand of ‘->’ has non-pointer type ‘ros::NodeHandle’
   ros::Subscriber ego_control_sub = ego_control_node_handle->subscribe("/your_controller_topic/vehicle_control", 1, [egoControlServer = std::move(egoControlServer)](const monodrive_msgs::VehicleControl &vehicle_control){
                                                            ^~
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp: In lambda function:
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:215:5: error: ‘egoControlServer’ is not captured
     egoControlServer->sendCommand(ApiMessage(777, EgoControl_ID, true, egoControl.dump()));
     ^~~~~~~~~~~~~~~~
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:209:164: note: the lambda has no capture-default
 o_control_node_handle->subscribe("/your_controller_topic/vehicle_control", 1, [egoControlServer = std::move(egoControlServer)](const monodrive_msgs::VehicleControl &vehicle_control){
                                                                                                                              ^
/home/bradnull/projects/monodrive-client/examples/ros/src/distributed/src/distributed.cpp:206:8: note: ‘<typeprefixerror>egoControlServer’ declared here
   auto egoControlServer = std::make_unique<Server>(primaryServer->GetSimulator()->getServerIp(), primaryServer->GetSimulator()->getServerPort());
        ^~~~~~~~~~~~~~~~
distributed/CMakeFiles/distributed_node.dir/build.make:62: recipe for target 'distributed/CMakeFiles/distributed_node.dir/src/distributed.cpp.o' failed
make[2]: *** [distributed/CMakeFiles/distributed_node.dir/src/distributed.cpp.o] Error 1
CMakeFiles/Makefile2:470: recipe for target 'distributed/CMakeFiles/distributed_node.dir/all' failed
make[1]: *** [distributed/CMakeFiles/distributed_node.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 88%] Linking CXX executable /home/bradnull/projects/monodrive-client/examples/ros/devel/lib/simulator_control/node
[ 88%] Built target simulator_control_node
[ 94%] Linking CXX executable /home/bradnull/projects/monodrive-client/examples/ros/devel/lib/vehicle_control/node
[ 94%] Built target vehicle_control_node
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j16 -l16" failed
```

File: monodrive/core/src/DistributedServer.h (hunk 1 above)
Comment (stephendb) [reply monodrive/core/src/DistributedServer.h]: name changed, also made inline

File: examples/cpp/distributed/distributed.cpp (hunk 2 above)
Comment (stephendb) [reply examples/cpp/distributed/distributed.cpp]: added

File: examples/ros/src/distributed/src/distributed.cpp (hunk 3 above)
Comment (stephendb) [reply examples/ros/src/distributed/src/distributed.cpp]: added

File: monodrive/core/src/Simulator.h (hunk 4 above)
Comment (stephendb) [reply monodrive/core/src/Simulator.h:60]: Yes, in the prior discussion I mentioned that the long term correction is to have Simulator allow multiple connections, hence configuration and connections are handled by simulator. Other threads that desire a connection get a connection etc.

**Comment by stephendb** [comment]
Fixed the link errors and applied changes based on comments.

**Pull request merged**